[BEAM-1981] Fix ApexTimerInternals serialization error.

tweise opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
R: @kennknowles @dhalperi

tweise commented:
```
Results :

Tests run: 221, Failures: 0, Errors: 0, Skipped: 3
```

kennknowles commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/utils/ApexTimerInternals.java:
```diff
@@ -0,0 +1,201 @@
+ * @param <K>
+ */
+@DefaultSerializer(JavaSerializer.class)
+public class ApexTimerInternals<K> implements TimerInternals, Serializable {
```
Could keep this package private if you leave it in the same directory, just to keep the API totally hidden from users.

tweise replied:
Moved the class.

kennknowles commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/utils/ApexTimerInternals.java:
```diff
@@ -0,0 +1,201 @@
+@DefaultSerializer(JavaSerializer.class)
+public class ApexTimerInternals<K> implements TimerInternals, Serializable {
+
+  private Map<Slice, Set<Slice>> activeTimers = new HashMap<>();
```
You may want a `NavigableSet` in case many windows become active on a key (I think this would require the watermark to stall for a bit).

Given the way you implement this - with the impetus for processing coming separately and just saving the timer data here - could you easily support cancel in the same manner as the direct runner? https://github.com/apache/beam/blob/master/runners/direct-java/src/main/java/org/apache/beam/runners/direct/WatermarkManager.java#L223

tweise replied:
You mean deleteTimer(..)? Yes, that would work in the same manner as I see it in the direct runner. Will take that up later though.

kennknowles commented on runners/apex/src/main/java/org/apache/beam/runners/apex/translation/utils/ApexTimerInternals.java:
```diff
@@ -0,0 +1,201 @@
+  @Override
+  public void setTimer(StateNamespace namespace, String timerId, Instant target,
+      TimeDomain timeDomain) {
+    TimerData timerData = TimerData.of(timerId, namespace, target, timeDomain);
```
I understand this is just a move of existing code. But here you enqueue timers for all time domains and below treat them all as event time timers. I think either you should make two sets/queues or else throw unsupported for processing time timers.

Is it correct that something like watermark update tuples will continue to cause the operator to inquire about ready timers? In that case the two queue solution is easy and it is actually OK to use `Instant.now()` since there's no distributed clock problem.

tweise replied:
Yes, the timers are currently processed when the input watermark is updated. Having separate sets will make sense because the processing timers need to be looked at irrespective of watermark change. I would prefer to address that as part of the other JIRA though. This here is to fix the serialization issue.

kennknowles commented on runners/apex/src/test/java/org/apache/beam/runners/apex/translation/utils/ApexTimerInternalsTest.java:
```diff
@@ -0,0 +1,56 @@
+public class ApexTimerInternalsTest {
+
+  @Test
+  public void testSerialization() {
```
I would suggest one or two smoke tests here for different time domains, since actually our validation suite doesn't really exercise this code much.

tweise replied:
Added test for event time timers.

tweise commented:
@kennknowles addressed comments. Processing time timer changes will be addressed as part of BEAM-2022, after this is merged. Want to get both fixed for the release.

kennknowles commented:
LGTM. Merging
